components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/add-security-questions.jsp

```diff
@@ -0,0 +1,157 @@
+
+<%@ page import="java.util.ArrayList" %>
+<%@ page import="java.util.Iterator" %>
+<%@ page import="java.util.HashMap" %>
+<%@ page import="java.util.List" %>
+<%@ page import="java.util.Map" %>
+
+<%
+    String BUNDLE = "org.wso2.carbon.identity.application.authentication.endpoint.i18n.Resources";
+    ResourceBundle resourceBundle = ResourceBundle.getBundle(BUNDLE, request.getLocale(), new
+            EncodedControl(StandardCharsets.UTF_8.toString()));
+    String urlData = request.getParameter("data");
+    // Extract the challege questions from the request and add them into an array
+    String[] questionSets = urlData.split("&");
+    // Hashmap to hold user's challenge questions
+    Map<String, List<ChallengeQuestion>> challengeQuestionMap = new HashMap<>();
+
+    for (String question : questionSets) {
+        String[] questionProperties = question.split("\\|");
+        // Construct a new ChallengeQuestion for each challenge question received from the request
+        ChallengeQuestion tempChallengeQuesion = new ChallengeQuestion();
+        tempChallengeQuesion.setQuestionSetId(questionProperties[0]);
+        tempChallengeQuesion.setQuestionId(questionProperties[1]);
+        tempChallengeQuesion.setQuestion(questionProperties[2]);
+        // Add the challenge question to the Hashmap
+        List<ChallengeQuestion> challengeQuestionList = challengeQuestionMap.get(questionProperties[0]);
+        if (challengeQuestionList == null) {
+            List<ChallengeQuestion> tempList = new ArrayList<>();
+            tempList.add(tempChallengeQuesion);
+            challengeQuestionMap.put(questionProperties[0], tempList);
+        } else {
+            challengeQuestionList.add(tempChallengeQuesion);
+        }
+    }
+%>
+
+<script type="text/javascript">
+    function validate() {
+        var isValid = new Boolean(true);
+        $("#profile input").each(function () {
+            if ($(this).val() == "") {
+                isValid = false;
+            }
+        })
+        if (isValid) {
+            document.getElementById("profile").submit();
+        }
+        else {
+            alert("Please enter the answers for challenge questions");
+        }
+    }
+</script>
+
+<!DOCTYPE html>
+<html lang="en">
+<html>
+<head>
+    <title><%=AuthenticationEndpointUtil.i18n(resourceBundle, "add.challenge.answers")%>
+    </title>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
+</head>
+
+<body>
+<div class="container">
+    <h1><%=AuthenticationEndpointUtil.i18n(resourceBundle, "answer.following.questions")%>
+    </h1>
+    <%
+        for (String list1 : challengeQuestionMap.keySet()) {
```

Comment:
Shall we have better names for the variables.

Reply:
fixed